### DISCUSSION TITLE
MacOS - zsh - local & global installations not working

### DISCUSSION THREAD

**Discussion by james-s-w-clark** [jdx/mise#703] [Q&A] [answered]
```
brew install rtx
rtx local erlang@25 && rtx local elixir@1.15.0-otp-25
rtx install
rtx use erlang@25 && rtx use elixir@1.15.0-otp-25

elixir -v 
zsh: command not found: elixir

rtx doctor
...
1 problem found:
rtx is not activated, run `rtx activate` for setup instructions

echo 'eval "$(~/bin/rtx activate zsh)"' >> ~/.zshrc
source ~/.zshrc
rtx doctor
...
1 problem found:
rtx is not activated, run `rtx activate` for setup instructions


rtx use --global node@18
...
node -v
zsh: command not found: node
```

My `~/.zshrc` has `eval "$(~/bin/rtx activate zsh)"` at the end, from `echo 'eval "$(~/bin/rtx activate zsh)"' >> ~/.zshrc`. `rtx doctor` is still not happy, as shown.

I see there is some stuff in the readme about using shims, but I thought it's better not to use these (`asdf` did, and they added ~100ms and that was a big motivator for `rtx`)?

--- 

Full `rtx doctor` output:

```
➜  ~ rtx doctor
rtx version:
  1.34.1 macos-x64 (unknown 2023-07-14)

build:
  Target: x86_64-apple-darwin
  Features: BREW
  Built: Fri, 14 Jul 2023 18:57:24 +0000
  Rust Version: rustc 1.70.0 (90c541806 2023-05-31) (built from a source tarball)
  Profile: release

shell:
  /bin/zsh
  zsh 5.9 (x86_64-apple-darwin22.0)

rtx environment variables:
  (none)

settings:
  {"always_keep_download": "false", "always_keep_install": "false", "asdf_compat": "false", "disable_default_shorthands": "false", "disable_tools": "[]", "experimental": "false", "jobs": "4", "legacy_version_file": "true", "legacy_version_file_disable_tools": "[]", "log_level": "INFO", "missing_runtime_behavior": "warn", "plugin_autoupdate_last_check_duration": "10080", "raw": "false", "trusted_config_paths": "[]", "verbose": "false"}

config files:
  /Users/myUser/.config/rtx/config.toml
  /Users/myUser/.tool-versions

plugins:
  elixir   https://github.com/asdf-vm/asdf-elixir.git#a4c42e1
  erlang   https://github.com/asdf-vm/asdf-erlang.git#127716e
  node     (core)
  python   (core)

toolset:
  elixir@1.15.0-otp-25, erlang@25, node@18

1 problem found:
rtx is not activated, run `rtx activate` for setup instructions
```

**Comment by james-s-w-clark** [comment]
This works!
```
curl https://rtx.pub/install.sh | sh

# as suggested after rtx shell install
echo -e "\neval \"\$(/Users/james/.local/share/rtx/bin/rtx activate -s zsh)\"" >> ~/.zshrc

➜  ~ source ~/.zshrc
[WARN] rtx::toolset: Tools not installed: elixir@1.15.0-otp-25, erlang@25, node@18.17.0 (install with: rtx install)
➜  ~ rtx install

rtx node@18.17.0 9.6.7                                                                                                                                                          ✓ 7s
➜  ~ rtx current
elixir 1.15.0-otp-25
erlang 25.3.2.4
node 18.17.0
➜  ~ elixir -v
Erlang/OTP 25 [erts-13.2.2.2] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit:ns]

Elixir 1.15.0 (compiled with Erlang/OTP 25)
```

So, I think there was some issue or user error with the homebrew setup.
This was much smoother.

**Comment by dfang** [comment]
when you install rtx via brew, rtx binary will install to `/usr/local/bin/rtx`.
so you need change `echo 'eval "$(~/bin/rtx activate zsh)"' >> ~/.zshrc` to  `echo 'eval "$(/usr/local/bin/rtx activate zsh)"' >> ~/.zshrc`

**Comment by jdx** [comment]
does `eval "$(~/bin/rtx activate zsh)"` not show an error message if the file doesn't exist?

**Reply by james-s-w-clark** [reply]
You're right:
```
➜  havvk git:(master) ✗ echo 'eval "$(~/bin/rtx activate zsh)"' >> ~/.zshrc
➜  havvk git:(master) ✗ source ~/.zshrc
/Users/james/.zshrc:122: no such file or directory: /Users/james/bin/rtx
```

```
echo 'eval "$(rtx activate)"' >> ~/.zshrc
source ~/.zshrc
elixir --version
# it works fine
```


---

I can't explain what the heck I was trying/saying there 😁 somehow I tried a bunch of stuff but nothing worked. Definitely PEBCAK!

FWIW, I did the brew install again; Brew doesn't suggest adding anything to ~/.zshrc (the `| sh` install does):

```
➜  havvk git:(master) ✗ brew install rtx
==> Fetching rtx
==> Downloading https://ghcr.io/v2/homebrew/core/rtx/manifests/1.34.1
Already downloaded: /Users/james/Library/Caches/Homebrew/downloads/e503f1a70dd858074cd14f457611dd941aa3b27d465ac74023b3305982bdb4c8--rtx-1.34.1.bottle_manifest.json
==> Downloading https://ghcr.io/v2/homebrew/core/rtx/blobs/sha256:49eca819a879985c2559d564774545d98a878ccb20251659988f5ffffada2629
Already downloaded: /Users/james/Library/Caches/Homebrew/downloads/475c49b14e2d45e858b84ab47c0c5d4fef1c57857c322f24efc37f2922fa643f--rtx--1.34.1.ventura.bottle.tar.gz
==> Pouring rtx--1.34.1.ventura.bottle.tar.gz
==> Caveats
zsh completions have been installed to:
  /usr/local/share/zsh/site-functions
==> Summary
🍺  /usr/local/Cellar/rtx/1.34.1: 11 files, 14.6MB
==> Running `brew cleanup rtx`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
[WARN] rtx::toolset: Tools not installed: erlang@25, elixir@1.15.0-otp-25, mill@0.11.0 (install with: rtx install)
```

I think that's what helped me get going smoothly with that install.

So now ~.zshrc has 
`eval "$(rtx activate zsh)"`
and `elixir --version` works

**Reply by james-s-w-clark** [reply]
BTW, it looks like there's a few ways of doing things in the readme:

- eval "$(rtx activate)"
- eval "$(~/bin/rtx activate bash)"
    - if the first one works:
      - why do we need to specify a path 
      - why specify the type shell to activate?  

---

Amazing tool though! I was really happy with the templating - I just wrote a blog on it 😁  https://idiosapps.github.io/garden/rtx-asdf/

The key interesting bit is getting versions from various bits of source code, to prevent drift:

```toml
[tools]
java = { version = "{{exec(command='head -1 docker/Dockerfile | cut -c 12- | tr : -')}}" }
mill = { version = "{{exec(command='echo $(cat .mill-version)')}}" }
scala = { version = "{{exec(command='grep -m 1 2.13 dependencies/Dependencies.sc | cut -c 33-39')}}" }
```

**Reply by james-s-w-clark** [reply]
I saw another two things, apologies for so many replies:

```
➜  ~ rtx doctor
1 problem found:
rtx is not activated, run `rtx activate` for setup instructions

➜  ~ rtx activate
export RTX_SHELL=zsh
...
if [[ -z "${chpwd_functions[(r)_rtx_hook]+1}" ]]; then
  chpwd_functions=( _rtx_hook ${chpwd_functions[@]} )
fi
```

`rtx activate` doesn't really give setup instructions - should `rtx doctor` point to the relevant bit(s) of readme when it sees that issue instead?

---

`rtx` or `rtx --help` say this for activate:
`activate     Initializes rtx in the current shell`
FWIW, `rtx activate` and then `elixir --version` (when ~/.zshrc has no `rtx activate stuff`) gives `zsh: command not found: elixir`. Should that be working, even if only for the current shell session?

If you think they're worthwhile issues, I can raise them later. The discussion got on a bit of a tangent as I tried to figure out why I got lost this morning with the brew rtx setup.

**Comment by dfang** [comment, marked as answer]
> BTW, it looks like there's a few ways of doing things in the readme:
> 
> * eval "$(rtx activate)"
> * eval "$(~/bin/rtx activate bash)"
>   
>   * if the first one works:
>     
>     * why do we need to specify a path
>     * why specify the type shell to activate?
> 

if  `rtx` is in your path, and you're using bash, just `eval "$(rtx activate bash)` is enough. if `rtx` not in your path, you have to use absolute path.

for different shell , `rtx activate SHELL_TYPE` generates different output.

**Comment by dfang** [comment]
@jdxcode I suggest `rtx doctor` add some information about where rtx keeps downloads, installs plugins and apps. by default it's `~/.local/share/rtx`, but it can be elsewhere if `RTX_DATA_DIR` is set.  if this is acceptable, maybe I can draft a pr.

**Comment by oskar-gmerek** [comment]
I also encountered this problem, and I wasted some time before finding this thread. I must admit that the thought of trying a different manager crossed my mind.

I believe it would be worthwhile to link this thread in the quickstart guide. Maybe it's better to simply change the quickstart commands from:

```sh
echo 'eval "$(~/bin/rtx activate bash)"' >> ~/.bashrc
echo 'eval "$(~/bin/rtx activate zsh)"' >> ~/.zshrc
echo '~/bin/rtx activate fish | source' >> ~/.config/fish/config.fish
```
to:
```sh
echo 'eval "$(rtx activate)"' >> ~/.bashrc
echo 'eval "$(rtx activate)"' >> ~/.zshrc
echo 'rtx activate | source' >> ~/.config/fish/config.fish
```
Wouldn't that possibly solve most of these types of issues?

I personally on `macOS 14` and I installed `rtx` via `homebrew`

**Reply by jdx** [reply]
In the quickstart you run the following commands:

```
$ curl https://rtx.pub/rtx-latest-macos-arm64 > ~/bin/rtx
$ chmod +x ~/bin/rtx
$ rtx --version
rtx 1.35.7
echo 'eval "$(~/bin/rtx activate bash)"' >> ~/.bashrc
```

If `~/bin` is not on PATH (which more likely than not it won't be), then this won't function if you didn't specify the full path.

**Reply by oskar-gmerek** [reply]
Oh yeah, I see this now. This was come from choosing a brew over curl. 

In any case, I think it would be worth pointing out somewhere that using homebrew, for example, will result in the need to adjust the rest of the instructions in quickstart.

Personally, I always use brev whenever possible. And I don't recall ever having problems completing quickstarts using it.

I understand that this is my mistake and perhaps a slight ignorance. However, simply mentioning it prominently could have helped avoid this problem.

rtx looks like a great tool, I hope it develops quickly and a huge community is built around it. Somehow making sure future users never experience similar problems might help at least a little.

All the best.
